packages/core/graph/src/AdjacencyList.js

```diff
@@ -0,0 +1,1210 @@
+      }
+      address = this.next(address);
+    }
+    return null;
+  }
+
+  from(edge: EdgeAddress): NodeId {
+    return toNodeId(this.data[edge + EdgeTypeMap.#FROM]);
+  }
+
+  to(edge: EdgeAddress): NodeId {
+    return toNodeId(this.data[edge + EdgeTypeMap.#TO]);
+  }
+
+  nextIn(edge: EdgeAddress): EdgeAddress | null {
+    return this.data[edge + EdgeTypeMap.#NEXT_IN] || null;
+  }
+
+  prevIn(edge: EdgeAddress): EdgeAddress | null {
+    return this.data[edge + EdgeTypeMap.#PREV_IN] || null;
+  }
+
+  linkIn(edge: EdgeAddress, next: EdgeAddress) {
+    this.data[edge + EdgeTypeMap.#NEXT_IN] = next;
+    this.data[next + EdgeTypeMap.#PREV_IN] = edge;
+  }
+
+  unlinkIn(edge: EdgeAddress) {
+    let next = this.nextIn(edge);
+    let prev = this.prevIn(edge);
+    this.data[edge + EdgeTypeMap.#NEXT_IN] = 0;
+    this.data[edge + EdgeTypeMap.#PREV_IN] = 0;
+    if (next !== null && prev !== null) {
+      this.data[prev + EdgeTypeMap.#NEXT_IN] = next;
+      this.data[next + EdgeTypeMap.#PREV_IN] = prev;
+    } else if (next !== null) {
+      this.data[next + EdgeTypeMap.#PREV_IN] = 0;
+    } else if (prev !== null) {
+      this.data[prev + EdgeTypeMap.#NEXT_IN] = 0;
+    }
+  }
+
+  nextOut(edge: EdgeAddress): EdgeAddress | null {
+    return this.data[edge + EdgeTypeMap.#NEXT_OUT] || null;
+  }
+
+  prevOut(edge: EdgeAddress): EdgeAddress | null {
+    return this.data[edge + EdgeTypeMap.#PREV_OUT] || null;
+  }
+
+  linkOut(edge: EdgeAddress, next: EdgeAddress) {
+    this.data[edge + EdgeTypeMap.#NEXT_OUT] = next;
+    this.data[next + EdgeTypeMap.#PREV_OUT] = edge;
+  }
+
+  unlinkOut(edge: EdgeAddress) {
+    let next = this.nextOut(edge);
+    let prev = this.prevOut(edge);
+    this.data[edge + EdgeTypeMap.#NEXT_OUT] = 0;
+    this.data[edge + EdgeTypeMap.#PREV_OUT] = 0;
+    if (next !== null && prev !== null) {
+      this.data[prev + EdgeTypeMap.#NEXT_OUT] = next;
+      this.data[next + EdgeTypeMap.#PREV_OUT] = prev;
+    } else if (next !== null) {
+      this.data[next + EdgeTypeMap.#PREV_OUT] = 0;
+    } else if (prev !== null) {
+      this.data[prev + EdgeTypeMap.#NEXT_OUT] = 0;
+    }
+  }
+
+  /** Create a hash of the edge connecting the `from` and `to` nodes.  */
+  hash(from: NodeId, to: NodeId, type: TEdgeType): EdgeHash {
+    // Each parameter is hashed by mixing its upper bits into its lower bits to
+    // increase the likelihood that a change to any bit of the input will vary
+    // the output widely. Then we do a series of prime multiplications and
+    // additions to combine the hashes into one value.
+    let hash = 17;
+    hash = hash * 37 + hash32shift((from: any));
+    hash = hash * 37 + hash32shift((to: any));
+    hash = hash * 37 + hash32shift((type: any));
+    // Finally, we map the hash to a value modulo the edge capacity.
+    hash %= this.capacity;
+    return hash;
+  }
+}
+
+// From https://gist.github.com/badboy/6267743#32-bit-mix-functions
+function hash32shift(key: number): number {
```

Comment:
This crazy bit mixing function improves the uniformity of our hash table a fair amount without adding any noticeable time to hashing. We found this in the [v8 Map hash utils](https://github.com/nodejs/node/blob/238104c531219db05e3421521c305404ce0c0cce/deps/v8/src/utils/utils.h#L213)